### PR TITLE
Add permissions model to key store

### PR DESCRIPTION
**Pull request by inorick**

### CODE REVIEW

File: heimlig/src/hsm/keystore.rs [hunk 1]
```diff
@@ -106,50 +134,117 @@ impl<const STORAGE_SIZE: usize, const NUM_KEYS: usize> KeyStore
     for MemoryKeyStore<STORAGE_SIZE, NUM_KEYS>
 {
     fn is_stored(&self, id: KeyId) -> bool {
-        match self.infos.iter().find(|key_info| key_info.id == id) {
+        match self
+            .layout
+            .iter()
+            .find(|key_layout| key_layout.info.id == id)
```
Comment (rpiasetskyi) [inline heimlig/src/hsm/keystore.rs]: Does it make sense to store only the sorted layout and use the binary search to improve the performance? https://doc.rust-lang.org/nightly/core/primitive.slice.html#method.binary_search

File: heimlig/src/hsm/keystore.rs [hunk 2]
```diff
@@ -106,50 +134,117 @@ impl<const STORAGE_SIZE: usize, const NUM_KEYS: usize> KeyStore
...
+        {
             None => false,
-            Some(key_info) => key_info.actual_size > 0,
+            Some(key_layout) => key_layout.actual_size > 0,
         }
     }
 
     fn size(&self, id: KeyId) -> Result<usize, Error> {
-        match self.infos.iter().find(|key_info| key_info.id == id) {
```
Comment (rpiasetskyi) [inline heimlig/src/hsm/keystore.rs:116]: Does it make sense to have some function that returns `Option<KeyLayout>` that can be used by `is_stored`, `export`, `size`, `delete` etc functions?

File: heimlig/src/hsm/keystore.rs [hunk 3]
```diff
@@ -106,50 +134,117 @@ impl<const STORAGE_SIZE: usize, const NUM_KEYS: usize> KeyStore
...
+                }
+                if key_layout.actual_size == 0 {
                     return Err(Error::KeyNotFound);
                 }
-                if dest.len() < key_info.actual_size {
+                let offset = key_layout.offset;
+                let size = key_layout.actual_size;
+                let key = &mut self.storage[offset..(offset + size)];
+                key.fill(0);
+                key_layout.actual_size = 0;
```
Comment (rpiasetskyi) [inline heimlig/src/hsm/keystore.rs:450]: After the key is deleted. should we mark it somewhere as such?

File: heimlig/src/client/api.rs [hunk 4]
```diff
@@ -9,12 +9,12 @@ pub struct Api<'data, Req: Sink<Request<'data>>, Resp: Stream<Item = Response<'d
     request_id_counter: RequestId,
 }
 
-#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```
Comment (rpiasetskyi) [inline heimlig/src/client/api.rs:12]: Were these reorderings expected? :)

File: heimlig/src/hsm/keystore.rs (hunk 2 above)
Comment (inorick) [reply heimlig/src/hsm/keystore.rs:116]: Yes, there is `get_key_layout` now.

File: heimlig/src/hsm/keystore.rs (hunk 3 above)
Comment (inorick) [reply heimlig/src/hsm/keystore.rs:450]: We currently use `key_layout.actual_size == 0` to determine this (see `is_stored`). I feel like an additional status flag would just duplicate state with `actual_size`.

File: heimlig/src/client/api.rs (hunk 4 above)
Comment (inorick) [reply heimlig/src/client/api.rs:12]: I wanted to sort them alphabetically.

File: heimlig/src/hsm/keystore.rs [hunk 5]
```diff
@@ -222,35 +245,36 @@ impl<const STORAGE_SIZE: usize, const MAX_KEYS: usize> MemoryKeyStore<STORAGE_SI
...
+            layout: SortedKeyStoreLayout::try_from(key_infos)?,
+        })
     }
 }
 
 impl<const STORAGE_SIZE: usize, const NUM_KEYS: usize> KeyStore
     for MemoryKeyStore<STORAGE_SIZE, NUM_KEYS>
 {
     fn get_key_info(&self, id: KeyId) -> Result<KeyInfo, Error> {
-        match self.get_key_layout(id) {
```
Comment (rpiasetskyi) [inline heimlig/src/hsm/keystore.rs]: Same question about the 
```
self.layout
            .get(id)
            .map(|key_layout| key_layout.info)
            .ok_or(Error::InvalidKeyId)
```

:see_no_evil:

File: heimlig/src/hsm/keystore.rs [hunk 6]
```diff
@@ -205,15 +230,13 @@ impl<const STORAGE_SIZE: usize, const MAX_KEYS: usize> MemoryKeyStore<STORAGE_SI
             return Err(Error::DuplicateIds);
         }
 
-        // Create new key store
-        let mut key_store = MemoryKeyStore {
-            storage: [0u8; STORAGE_SIZE],
-            layout: Default::default(),
+        // Create new sorted key layout
+        let mut ret = SortedKeyStoreLayout {
+            inner: Default::default(),
```
[the remark this answers is not in the document]
Comment (inorick) [reply heimlig/src/hsm/keystore.rs:235]: I ended up using `Self::default()` and simplified the loop a little :+1:

**Pull request merged**